src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2024,147 +1981,192 @@ else if (compareAndSetCtl(c, c) && casRunState(e, e | STOP)) {
-        if (w != null) {
-            w.phase = p;                         // deactivate
-            long np = active & LMASK, pc = ctl;  // try to enqueue
-            long qc = np | ((pc - RC_UNIT) & UMASK);
+    private int deactivate(WorkQueue w, int r, int phase) {
+        int p = phase | IDLE;
+        if (w != null) {                         // currently always true
+            w.phase = p;
+            int active = phase + (IDLE << 1), ac;
+            long pc = ctl, qc = (active & LMASK) | ((pc - RC_UNIT) & UMASK);
             w.stackPred = (int)pc;               // set ctl stack link
-            if (pc != (pc = compareAndExchangeCtl(pc, qc))) {
-                qc = np | ((pc - RC_UNIT) & UMASK);
-                w.stackPred = (int)pc;           // retry once
-                if (pc != (pc = compareAndExchangeCtl(pc, qc)))
-                    p = w.phase = phase;         // back out
-            }
-            if (p != phase && ((e = runState) & STOP) == 0 &&
-                (!(quiet = (qc & RC_MASK) <= 0L) || (e & SHUTDOWN) == 0 ||
-                 !(quiet = quiescent()) || (runState & STOP) == 0)) {
-                long deadline = 0L;              // not terminating
-                if (quiet) {                     // use timeout if trimmable
-                    int nt = (short)(qc >>> TC_SHIFT);
-                    long delay = keepAlive;      // scale if not at target
-                    if (nt != (nt = Math.max(nt, parallelism)) && nt > 0)
-                        delay = Math.max(TIMEOUT_SLOP, delay / nt);
-                    if ((deadline = delay + System.currentTimeMillis()) == 0L)
-                        deadline = 1L;           // avoid zero
-                }
-                boolean release = quiet;
-                WorkQueue[] qs = queues;         // recheck queues
-                int n = (qs == null) ? 0 : qs.length;
-                for (int l = -n, j = active; l < n; ++l, ++j) {
-                    WorkQueue q; ForkJoinTask<?>[] a; int cap;
-                    if ((p = w.phase) == active) // interleave signal checks
-                        break;
-                    if ((q = qs[j & (n - 1)]) != null &&
-                        (a = q.array) != null && (cap = a.length) > 0 &&
-                        a[q.base & (cap - 1)] != null) {
-                        if (release && qc == ctl && compareAndSetCtl(qc, pc)) {
-                            p = w.phase = active;
-                            break;               // possible missed signal
+            if ((runState & STOP) != 0L || !compareAndSetCtl(pc, qc))
+                p = w.phase = phase;             // back out
+            else if ((ac = (short)(qc >>> RC_SHIFT)) > 0 ||
+                     quiescent() <= 0) {         // check quiescent termination
+                WorkQueue[] qs;
+                int spins = ac + ((short)(qc >>> TC_SHIFT) << 1) + SPIN_WAITS;
+                while (((p = w.phase) & IDLE) != 0 && --spins != 0)
```

Comment:
Might be worth checking `--spins >= 0` to avoid some pathological case where SPIN_WAITS is set to 0?